Recover lost broker-info.subm

Jaanki opened the pull request:
Since there are multiple places where a broker installation check on a cluster is needed in multiple places (show brokers, uninstall and recover-brokerinfo (soon)), add broker field as part of the clusterinfo and a function that checks if the cluster has Broker installed.

This simplies code at multiple places.

Depends on: https://github.com/submariner-io/submariner-operator/pull/2432

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

skitt commented on pkg/cluster/info.go:
```diff
@@ -40,12 +39,13 @@ type Info struct {
 }
 
-func NewInfo(clusterName string, config *rest.Config) (*Info, error) {
+func NewInfo(clusterName, brokerNamespace string, config *rest.Config) (*Info, error) {
```
The `brokerNamespace` parameter is always `""`, let’s drop it for now and rely on the constant.

Jaanki replied:
For, `recover-brokerinfo` the namespace might be different and hence have added this parameter here.

skitt replied:
In this PR, it can’t be; you always pass in `""`.

skitt reviewed (changes requested):
The PR description mentions that this simplifies code in a number of places, the PR should demonstrate this.

skitt commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/submariner-io/subctl
 
-go 1.19
+go 1.18
```
Please don’t revert to 1.18.

Jaanki replied:
I did not. Don't know how this ended up changing.

skitt commented on pkg/cluster/info.go:
```diff
@@ -174,3 +174,16 @@ func (c *Info) OperatorNamespace() string {
 	return constants.OperatorNamespace
 }
+
+func (c *Info) GetBrokers(namespace string) (*v1alpha1.BrokerList, bool, error) {
```
There’s no need for a boolean here; if the function can’t return any results, make it return `nil`.

```suggestion
func (c *Info) GetBrokers(namespace string) (*v1alpha1.BrokerList, error) {
```

tpantelis replied:
Also for convenience, we can return `[]v1alpha1.Broker` since `BrokerList.Items` is what callers will access anyway (it's the only field in `BrokerList`).

Jaanki replied:
> There’s no need for a boolean here; if the function can’t return any results, make it return `nil`.

The bool is returned so that the calling function can use `if found` logic. Ofcourse, we can do `if len(Brokers)` logic but I find the former more readable.

skitt replied:
You’re returning the same information in two different forms, which is generally speaking not a good idea — if the behaviour changes in future, you can end up with diverging handling in different code paths (depending on whether the code uses `len(Brokers)` or `found`).

tpantelis replied:
The caller has to check `len(brokers)` anyway. For the case where there's no error returned, ie `return brokerList, true, nil`,  the `brokerList` will contain an empty `Items` array if no brokers exist so retuning a `found` bool doesn't really add any value.

skitt commented on pkg/cluster/info.go:
```diff
@@ -174,3 +174,16 @@ func (c *Info) OperatorNamespace() string {
+	brokerList := &v1alpha1.BrokerList{}
+	err := c.ClientProducer.ForGeneral().List(context.TODO(), brokerList, controllerClient.InNamespace(namespace))
+
+	if err == nil {
+		return brokerList, true, nil
+	} else if !apierrors.IsNotFound(err) && !meta.IsNoMatchError(err) {
+		return nil, false, errors.Wrapf(err, "error retrieving Broker in namespace %s", namespace)
+	}
+
+	return brokerList, false, nil
```
The idiomatic pattern in Go is to look for non-nil `err`:

```suggestion
        if err != nil && !apierrors.IsNotFound(err) && !meta.IsNoMatchError(err) {
            return nil, errors.Wrapf(err, "error retrieving Broker in namespace %s", namespace)
        }
        
        return brokerList, nil
```

Jaanki replied:
Well, I just copied what is done for Submariner (https://github.com/submariner-io/subctl/blob/devel/pkg/cluster/info.go#L68-L72) and Service discovery.

skitt replied:
The usage is different: you’re returning in the `err == nil` branch, the code in `NewInfo` is filling in a value in the `err == nil` branch.

tpantelis commented on internal/show/brokers.go:
```diff
@@ -19,59 +19,53 @@ limitations under the License.
-			brokers[i].Spec.DefaultGlobalnetClusterSize,
-			brokers[i].Spec.DefaultCustomDomains,
-		)
+	if err != nil {
+		return status.Error(err, "error retrieving broker")
```
The error should be checked first, ie above L32 above.

tpantelis commented on internal/show/brokers.go:
```diff
@@ -70,8 +62,8 @@ func Brokers(clusterInfo *cluster.Info, _ string, status reporter.Interface) err
 
-	status.End()
 	printer.Print()
+	status.End()
```
Why was this moved? Actually either way this call is unnecessary b/c there's no status started at this point so we can remove it.

tpantelis commented on pkg/cluster/info.go:
```diff
@@ -59,6 +60,15 @@ func NewInfo(clusterName string, config *rest.Config) (*Info, error) {
+		info.Broker = brokers.Items
+	} else if !apierrors.IsNotFound(err) && !meta.IsNoMatchError(err) {
+		return nil, errors.Wrapf(err, "error retrieving Broker")
+	}
```
I think we should make this on-demand (as you originally had), ie function `GetBrokers`,  to reduce overhead as this is only needed for one command.

Jaanki replied:
Not one. This is used by `recover-brokerinfo` extensively ([first time](https://github.com/submariner-io/subctl/pull/500/files#diff-ae7e6105a69a92b24a4c33457e32203a26eab3dd52a5fc97b6e5a98c7e41d5eeR48), [second time](https://github.com/submariner-io/subctl/pull/500/files#diff-ae7e6105a69a92b24a4c33457e32203a26eab3dd52a5fc97b6e5a98c7e41d5eeR80) and [third time](https://github.com/submariner-io/subctl/pull/493/files#diff-3489601ae060d1918b959a7c59ae5a65d8b66f4af0c89ed222f99092ab4cc64dR35)). This make `recover-brokerinfo` logic easier. It can also be used by `uninstall` simplifying [findBrokerNamespace](https://github.com/submariner-io/subctl/blob/devel/pkg/uninstall/all.go#L324)

tpantelis replied:
yes but in those places the brokers can be still be retrieved on-demand via a function  and stored locally, ie they don't need to be retrieved up front and stored globally in the `Info` object.

tpantelis commented on pkg/cluster/info.go:
```diff
@@ -174,3 +175,20 @@ func (c *Info) OperatorNamespace() string {
+	}, broker)
+
+	if err == nil {
+		c.Broker = broker
```
This is stored but not used.

Jaanki committed:
Recover broker-info.subm file

This PR does the following:

1. Adds a `subctl recover-broker-info` command to recover
   broker-info.subm file from an existing cluster setup
2. Adds logic to collect necessary information from Broker and
   Submariner clusters to populate broker-info.subm file
3. Add system tests for various scenarios

Epic: https://github.com/submariner-io/enhancements/issues/143

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
```
Let's name the file after the actual command, ie _recover_broker_info.go_.

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+
+// recoverBrokerInfo represents the reconstruct command.
+var recoverBrokerInfo = &cobra.Command{
+	Use:   "recover-brokerinfo",
```
```suggestion
	Use:   "recover-broker-info",
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+// recoverBrokerInfo represents the reconstruct command.
+var recoverBrokerInfo = &cobra.Command{
+	Use:   "recover-brokerinfo",
+	Short: "Recovers broker-info.subm file from the installed Broker",
```
```suggestion
	Short: "Recovers the broker-info.subm file from the installed Broker",
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+	rootCmd.AddCommand(recoverBrokerInfo)
+}
+
+func addReconstructFlags() {
```
`addRecoverBrokerInfoFlags` but you don't really need a function here, just inline on L56.

Jaanki replied:
yes. Over the time, flags were removed from this function. This was a left over.

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+	// if --brokerconfig flag provided, get the broker and proceed to get Submariner
+	contextFound, err := recoverRestConfigProducer.RunOnSelectedPrefixedContext("broker",
+		func(brokerCluster *cluster.Info, brokerNamespace string, status reporter.Interface) error {
+			status.Start("Checking if Broker is installed on cluster %s in %s namespace", brokerCluster.Name, brokerNamespace)
```
```suggestion
			status.Start("Checking if the Broker is installed on cluster %q in namespace %q", brokerCluster.Name, brokerNamespace)
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+		func(brokerCluster *cluster.Info, brokerNamespace string, status reporter.Interface) error {
+			status.Start("Checking if Broker is installed on cluster %s in %s namespace", brokerCluster.Name, brokerNamespace)
+			defer status.End()
+			err := errors.New("error: ")
```
This is unused.

```suggestion
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+			if err != nil {
+				return err
+			}
+			if found {
```
```suggestion
			if err != nil {
				status.Error(err, "error retrieving the Broker information")
			}
			
			if !found {
			    return status.Error(errors.New("the Broker is not installed on the specified cluster. Please try again with another cluster"), "")
			}
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+					status.Success("Submariner found")
+					fmt.Printf("getting details")
+					return broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, status, ipsecFile)
+				}
```
```suggestion
				status.End()

				if brokerCluster.Submariner != nil {
					status.Success("Found Submariner installed on the same cluster")
					return broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, status, ipsecFile)
				}
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+					fmt.Printf("getting details")
+					return broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, status, ipsecFile)
+				}
+				status.Warning("Submariner is not installed on this cluster. Checking if Submariner is installed on a different cluster")
```
This really shouldn't be a warning. Let's start a new status sequence.

```suggestion
				status.Start("Submariner is not installed on this cluster. Checking if Submariner is installed on a different cluster")
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+
+						return status.Error(err, "submariner is not installed on this cluster. " +
+							"Please specify the cluster where Submariner is installed by using `--kubeconfig` or `--kubecontext` flag")
+					}, status)
```
```suggestion
						return status.Error(err, "Submariner is not installed on cluster %q. " +
							"Please specify the cluster where Submariner is installed by using the `--kubeconfig` or `--kubecontext` flag", submCluster.Name))
					}, status)
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+					func(submCluster *cluster.Info, namespace string, status reporter.Interface) error {
+						if submCluster.Submariner != nil {
+							status.Success("Submariner is installed on cluster %s", submCluster.Name)
+							fmt.Printf("getting details")
```
```suggestion
							status.Success("Submariner is installed on cluster %q", submCluster.Name)
							status.End()
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+		}, status)
+
+	// if --brokerconfig not provided, search for broker on current context and proceed to get Submariner
+	if !contextFound && err == nil {
```
Avoids all the indentation.

```suggestion
	if contextFound || err != nil {
	    return err
	}
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+					return nil
+				}
+
+				status.Success("Found Broker installed on the cluster")
```
```suggestion
				status.Start("Checking if the Broker is installed on cluster %q in namespace %q", clusterInfo.Name, namespace)
								
				found, err := clusterInfo.HasBroker(namespace)
				if err != nil {
					status.Error(err, "error retrieving the Broker information")
				}
				
				if !found {
				    return status.Error(errors.New("the Broker is not installed on the specified cluster. Please try again with another cluster"), "")
				}
				
				status.End()
```

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+				err = clusterInfo.ClientProducer.ForGeneral().List(context.TODO(), endpoints,
+					controllerClient.InNamespace(metav1.NamespaceAll))
+				if err != nil {
+					if apierrors.IsNotFound(err) {
```
`List` does not return an `IsNotFound` error if no items are found. You must check `if len(endpoints.Items) == 0`.

tpantelis commented on cmd/subctl/reconstruct.go:
```diff
@@ -0,0 +1,154 @@
+
+					status.Warning("Submariner is not installed on this cluster. " +
+						"Please specify the cluster where Submariner is installed via --kubeconfig or --kubecontext flag")
+				}
```
I'm not really clear on the logic here. It seems to me if the broker is installed then check if submariner is installed via `clusterInfo.Submariner != nil `as is done above. If not then check if there's no endpoints.

```
// broker found

if clusterInfo.Submariner != nil {
    status.Success("Found Submariner installed on the same cluster")
    return broker.ReconstructData(clusterInfo, clusterInfo, namespace, status, ipsecFile)
}

status.Start("Submariner is not installed on cluster %q. Checking if there are any joined Submariner clusters", clusterInfo.Name)

endpoints := &submarinerv1.EndpointList{}
err = clusterInfo.ClientProducer.ForGeneral().List(context.TODO(), endpoints,
    controllerClient.InNamespace(metav1.NamespaceAll))
if err != nil {
    return status.Error(err, "error retrieving Submariner endpoints")
}

if len(endpoints.Items) > 0 {
    return status.Error(fmt.Errorf("%d joined Submariner clusters were detected. The IPSec PSK key can only be recovered from a joined Submariner installation. Please specify a cluster where Submariner is installed via the --kubeconfig or --kubecontext flag in addition to the Broker cluster",  len(endpoints.Items)), "")
}

status.Success("No joined Submariner clusters were detected. ")
status.End()

return broker.ReconstructData(clusterInfo, clusterInfo, namespace, status, ipsecFile)
```

Jaanki replied:
The logic has been updated since this. PLease take a look again.

tpantelis commented on pkg/cluster/info.go:
```diff
@@ -174,3 +175,20 @@ func (c *Info) OperatorNamespace() string {
 	return constants.OperatorNamespace
 }
+
+func (c *Info) HasBroker(namespace string) (bool, error) {
```
The function name indicates it checks for existence but it also has the hidden side-effect of setting the `Broker` field which I don't think is proper. I think it's cleaner to return the `Broker` instance instead.

```suggestion
func (c *Info) GetBroker(namespace string) (*v1alpha1.Broker, bool, error) {
```

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+		return status.Error(err, "error getting broker client secret")
+	}
+
+	if brokerCluster.Broker != nil {
```
Have the caller pass in the `Broker` instance to the function. Also it shouldn't be nil at this point.

Jaanki replied:
Broker cluster is needed to retreive Broker URL at L25 above.

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+		data.CustomDomains = &brokerCluster.Broker.Spec.DefaultCustomDomains
+	} else if !apierrors.IsNotFound(err) && !meta.IsNoMatchError(err) {
+		return status.Error(err, "error retrieving Broker")
+	}
```
We already know `err` is nil at this point based on L31.

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+				Name: ipsecPSKSecretName,
+			},
+			Data: map[string][]byte{"psk": []byte(submCluster.Submariner.Spec.CeIPSecPSK)},
+		}
```
We really don't need the `submCluster`  - simplify the function by just passing in `Submariner`.

```suggestion
	if submariner != nil {
		status.Success("Retrieved the IPSec PSK secret from the Submariner installation")
		data.IPSecPSK = &v1.Secret{
			ObjectMeta: metav1.ObjectMeta{
				Name: ipsecPSKSecretName,
			},
			Data: map[string][]byte{"psk": []byte(submariner.Spec.CeIPSecPSK)},
		}
```

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+		ipsecData, err := ReadInfoFromFile(ipsecSubmFile)
+		if err != nil {
+			return errors.Wrapf(err, "error importing IPsec PSK secret from file %q", ipsecSubmFile)
+		}
```
```suggestion
		ipsecData, err := ReadInfoFromFile(ipsecSubmFile)
		if err != nil {
			return errors.Wrapf(err, "error importing IPsec PSK secret from file %q", ipsecSubmFile)
		}
		
		status.Success("Retrieved the IPSec PSK from the file %q", ipsecSubmFile)
```

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+		data.IPSecPSK, err = newIPSECPSKSecret()
+		if err != nil {
+			return status.Error(err, "error generating new IPSec PSK secret")
+		}
```
```suggestion
		data.IPSecPSK, err = newIPSECPSKSecret()
		if err != nil {
			return status.Error(err, "error generating new IPSec PSK secret")
		}
		
		status.Success("Generated a new IPSec PSK secret")
```

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+func ReconstructData(
+	brokerCluster, submCluster *cluster.Info, namespace string, status reporter.Interface,
+	ipsecSubmFile string) error {
+	status.Start("Retrieving data to reconstruct broker-info.subm")
```
```suggestion
	status.Start("Retrieving data to reconstruct the broker-info.subm")
```

tpantelis commented on pkg/broker/reconstruct.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+
+	status.Success("Successfully retrieved the data. Writing it to broker-info.subm")
+
+	err = data.writeToFile("broker-info.subm")
+
+	return status.Error(err, "error reconstructing broker-info.subm")
```
```suggestion
        status.End()
	status.Success("Writing the broker-info.subm file")
	defer status.End()

	err = data.writeToFile("broker-info.subm")

	return status.Error(err, "")
```

tpantelis reviewed (changes requested):
It would be good to add system tests to run the command with the various flag combinations and run join again afterwards.

tpantelis commented:
The PR summary also needs to be updated.

tpantelis commented on pkg/cluster/info.go:
```diff
@@ -220,3 +222,36 @@ func MergeImageOverrides(imageOverrides map[string]string, localImageOverrides [
+		context.TODO(), controllerClient.ObjectKey{
+			Namespace: namespace,
+			Name:      brokercr.Name,
+		}, broker)
```
To followup on the discussion re: moving this to `NewInfo`, it's problematic b/c we need the proper broker namespace which wouldn't be available in `NewInfo` for all commands up front.

This function should return `Broker`:

```
func (c *Info) GetBroker(namespace string) (*v1alpha1.Broker, error) {
   broker := &v1alpha1.Broker{}
    err := c.ClientProducer.ForGeneral().Get(
		context.TODO(), controllerClient.ObjectKey{
			Namespace: namespace,
			Name:      brokercr.Name,
		}, broker)

    return broker, err
}
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+var recoverRestConfigProducer = restconfig.NewProducer().WithPrefixedContext("broker").WithDefaultNamespace(constants.DefaultBrokerNamespace).
+	WithPrefixedNamespace("broker", constants.DefaultBrokerNamespace)
+
+var ipsecFile string
```
Let's store this in a struct as we've done with other commands:

```suggestion
recoverBrokerInfoConfig struct {
    ipsecFile string
}
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+
+func init() {
+	recoverBrokerInfo.PersistentFlags().StringVar(&ipsecFile, "ipsec-psk-from", "",
+		"import IPsec PSK from existing submariner broker file, like broker-info.subm")
```
This option doesn't seem useful.  If the user had an existing submariner broker file then they wouldn't need this command 😄

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+	// if --brokerconfig flag provided, get the broker and proceed to get Submariner
+	contextFound, err := recoverRestConfigProducer.RunOnSelectedPrefixedContext("broker",
+		func(brokerCluster *cluster.Info, brokerNamespace string, status reporter.Interface) error {
+			status.Start("Checking if Broker is installed on the specified cluster %s in %s namespace", brokerCluster.Name,
```
```suggestion
			status.Start("Checking if the Broker is installed on cluster %q in namespace %q", brokerCluster.Name, brokerNamespace", brokerCluster.Name,
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+			if err != nil {
+				return status.Error(err, "error getting Broker")
+			}
+			if found {
```
Return early here to avoid the excessive indentation.

```suggestion
			if !found {
			    return status.Error(errors.New("the Broker is not installed on the specified cluster. Please try again with another cluster"), "")
			}
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+					for i := range clusters {
+						if clusters[i].Spec.ClusterID == brokerCluster.Name {
+							status.Success("Submariner found")
+							return broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, status, ipsecFile)
```
We don't need to list the `Cluster` resources to check if submariner is installed - we can simply check `if brokerCluster.Submariner != nil` - that's what `ReconstructData` checks as well.  Also `brokerCluster.Name` is obtained from the kube config so I'm not sure it's guaranteed to match our cluster ID.

tpantelis replied:
As explained elsewhere, the purpose of searching for the local `Cluster` resource is to handle the case where there are multiple brokers installed, in which case we need to determine if the `Submariner` resource is actually joined to the broker to be recovered. What distinguishes multiple brokers is their namespaces.  So we can simply check if `brokerCluster.Submariner.Spec.BrokerK8sRemoteNamespace == brokerNamespace`.

Jaanki replied:
This checks the case where there are multiple clusters, each with broker and Submariner installed.

cluster1                    cluster2
Broker1                    Broker2
Submariner1          Submariner2

**Submariner2 is joined with Broker1. But Submariner1 is not joined with Broker1.**

When running recover command on cluster1 with context = cluster1, it will find both Broker1 and Submariner1. But since Submariner1 is  **NOT** connected to Broker1, it is the wrong Submariner instance for us and hence should not be used. Hence the command will fail saying that "a Submariner cluster" is joined to this Broker but we are not interested in this one. Please specify correct Submariner instance.

tpantelis replied:
right so this is the case where the user specified `--brokercontext cluster1`.  In your example, we would find Submariner1 but we need to check if it's joined to the local Broker1. You're searching for the `Cluster` resource but across all namespaces which is problematic - you really should only search the broker namespace. I suggested we check if `brokerCluster.Submariner.Spec.BrokerK8sRemoteNamespace == brokerNamespace` but that's not correct.  I think all we need to do is look up the `Cluster` resource associated with Submariner1, ie

```
cluster := &submarinerv1.Cluster{}
err := c.ClientProducer.ForGeneral().Get(context.TODO(), controllerClient.ObjectKey{
        Namespace: brokerNamespace,
	Name:      brokerCluster.Submariner.Spec.ClusterID,
     }, cluster)
```

If not found then we'd check the cluster associated with `--context` via `RunOnSelectedContext`.  It could be cluster2 or cluster1, either way we need to do the same `Cluster` lookup  as above to ensure it's joined to Broker1.

mkolesnik replied:
@tpantelis I think the whole confusion can be avoided and the UX can be much simpler - https://github.com/submariner-io/subctl/pull/467#discussion_r1103819587

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+							status.Success("Submariner found")
+							return broker.ReconstructData(clusterInfo, clusterInfo, namespace, status, ipsecFile)
+						}
+					}
```
If submariner isn't installed on the specified cluster it will fall through to L143 and generate a new PSK. We need to fail in this case.

Jaanki replied:
Logic is changed.This comment no longer applies.

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,148 @@
+			}, status)
+	}
+	return err
+}
```
There's some code duplication in the sections above. We can refactor and simplify a bit:

```
func recoverBroker(status reporter.Interface) error {
    contextFound, err := recoverRestConfigProducer.RunOnSelectedPrefixedContext("broker",
	func(brokerCluster *cluster.Info, brokerNamespace string, status reporter.Interface) error {
	    brokerObj, err := getBroker(brokerCluster, brokerNamespace, "Please try again with another cluster", status)
	    if err != nil {
	        return err
	    }

	    ok, err := tryToRecoverFromBroker(brokerCluster, brokerNamespace, brokerObj, status)
	    if ok || err != nil {
	        return err
	    }

	    return recoverRestConfigProducer.RunOnSelectedContext(
	        func(submCluster *cluster.Info, namespace string, status reporter.Interface) error {
		    if submCluster.Submariner != nil {
		        status.Success("Found Submariner installed on cluster %q", submCluster.Name)
			return broker.ReconstructData(brokerCluster, submCluster, brokerNamespace, brokerObj, status, ipsecFile)
		    }

		    return status.Error(fmt.Errorf("submariner is not installed on the Broker cluster %q. "+
			"Please specify the cluster where Submariner is installed via the `--kubeconfig` or `--kubecontext` flag"+
			submCluster.Name), "")
		}, status)
        }, status)

    if contextFound || err != nil {
        return err
    }

    return recoverRestConfigProducer.RunOnSelectedContext(
        func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error {
	    brokerObj, err := getBroker(clusterInfo, namespace, "Please specify the cluster where the Broker is installed via the "+
				"`--brokerconfig` or `--brokercontext` flag and the namespace via the --brokernamespace flag", status)
	    if err != nil {
                return err
	    }

	    ok, err := tryToRecoverFromBroker(clusterInfo, namespace, brokerObj, status)
	    if ok || err != nil {
	        return err
	    }

	    return status.Error(fmt.Errorf("submariner is not installed on the Broker cluster %q. "+
				"Please specify the cluster where Submariner is installed via the `--kubeconfig` or `--kubecontext` flag "+
				"and the cluster where the Broker is installed via the `--brokerconfig` or `--brokercontext` flag",
				clusterInfo.Name), "")
    }, status)
}

func tryToRecoverFromBroker(brokerCluster *cluster.Info, brokerNamespace string, brokerObj *v1alpha1.Broker,
    status reporter.Interface) (bool, error) {
    status.Start("Checking if there are any joined Submariner clusters")

    clusters, err := brokerCluster.GetJoinedClusters()
    if err != nil {
        return false, status.Error(err, "")
    }

    status.End()

    if len(clusters) == 0 {
        status.Success("No joined Submariner clusters were detected")
	return true, broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, brokerObj, status, ipsecFile)
    }

    if brokerCluster.Submariner != nil {
        status.Success("Found Submariner installed on the same cluster")
	return true, broker.ReconstructData(brokerCluster, brokerCluster, brokerNamespace, brokerObj, status, ipsecFile)
    }

    return false, nil
}

func getBroker(clusterInfo *cluster.Info, namespace string, notFoundMsg string, status reporter.Interface) (*v1alpha1.Broker, error) {
    status.Start("Checking if the Broker is installed on cluster %q in namespace %q", clusterInfo.Name, namespace)

    brokerObj, err := clusterInfo.GetBroker(namespace)
    if apierrors.IsNotFound(err) {
        return nil, status.Error(fmt.Errorf("the Broker is not installed on the specified cluster. %s", notFoundMsg), "")
    }

    if err != nil {
        return nil, status.Error(err, "")
    }

    status.End()

    return brokerObj, nil
}
```

tpantelis commented on pkg/cluster/info.go:
```diff
@@ -220,3 +222,36 @@ func MergeImageOverrides(imageOverrides map[string]string, localImageOverrides [
+	return false, nil
+}
+
+func (c *Info) GetJoinedClusters() ([]submarinerv1.Cluster, error) {
```
This function simply lists the `Clusters` so the term "Joined" is extraneous. 

We really should pass in the namespace to search, otherwise searching `NamespaceAll` would return duplicates if both the broker and submariner are installed.

Also checking for `IsNotFound` and `IsNoMatchError` below aren't necessary.

This function then becomes:

```
func (c *Info) GetClusters(namespace string) ([]submarinerv1.Cluster, error) {
    clusters := &submarinerv1.ClusterList{}
    err := c.ClientProducer.ForGeneral().List(context.TODO(), clusters, controllerClient.InNamespace(namespace))
    if err != nil {
        return nil, err
    }

    return clusters.Items, nil
}
```

Jaanki replied:
well, since we interested in just the Submariner installs, namespace could be hard-coded to `submariner-operator`

tpantelis replied:
No, we need to search the specified broker namespace. But, as per [my comment](https://github.com/submariner-io/subctl/pull/467#discussion_r1100542691) here, we can simply Get the desired `Cluster` so this function isn't needed.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -191,3 +192,14 @@ _subctl cloud prepare generic --kubecontext cluster1 && exit 1
 _subctl uninstall -y --context cluster2
 _subctl uninstall -y --kubeconfig "${KUBECONFIGS_DIR}"/kind-config-cluster1
+
+sleep 30s # Give time for uninstall to finish
```
It's best to avoid hard-coded sleeps. We should wait for the broker and/or submariner namespaces are deleted.

That said, perhaps we can move some of the tests above the uninstall tests on L193 and simplify to avoid uninstalling and re-installing, which may significantly increase the run time.   So prior to L193, we can test the cases where submariner is on same cluster as the broker. eg:

Expect success:

```
_subctl recover-broker-info --namespace submariner-k8s-broker --context cluster1
_subctl recover-broker-info --brokercontext cluster1
_subctl recover-broker-info --brokernamespace submariner-k8s-broker --brokercontext cluster1
```

To verify, we can save off the original `broker-info.subm` and compare with the recovered one.

Expect failures:

```
_subctl recover-broker-info --namespace non-existent --context cluster1
_subctl recover-broker-info --namespace submariner-k8s-broker --context cluster2
_subctl recover-broker-info --brokernamespace non-existent --brokercontext cluster1
_subctl recover-broker-info --brokernamespace submariner-k8s-broker --brokercontext cluster2
```

For these, verify a non-zero exit code.

After both clusters are uninstalled on L194, we can re-deploy the broker on cluster1 and test the case where there's no joined clusters and it generates a new IPsec secret:

```
_subctl recover-brokerinfo  --brokercontext cluster1
_subctl recover-broker-info --namespace submariner-k8s-broker --context cluster1
```

For the broker-only install case, we can join cluster2:

`_subctl recover-broker-info --brokercontext cluster1 --context cluster2`

skitt replied:
I think there’s value in having a well-identified `subctl uninstall` test; and if the `recover-brokerinfo` tests rely on `uninstall`, then the latter should be tested first, so that a breakage in `uninstall` doesn’t result in wasted time testing `recover-brokerinfo`.

Avoiding `sleep 30` would be great, but I’m not sure there’s a good way of doing that. `kubectl` can be asked to wait for certain states, perhaps that would work, if there’s an object which gives a reliable indication that the uninstallation is complete.

Jaanki replied:
> It's best to avoid hard-coded sleeps. We should wait for the broker and/or submariner namespaces are deleted.
> 
> That said, perhaps we can move some of the tests above the uninstall tests on L193 and simplify to avoid uninstalling and re-installing, which may significantly increase the run time. So prior to L193, we can test the cases where submariner is on same cluster as the broker. eg:
> 
> Expect success:
> 
> ```
> _subctl recover-broker-info --namespace submariner-k8s-broker --context cluster1
> _subctl recover-broker-info --brokercontext cluster1
> _subctl recover-broker-info --brokernamespace submariner-k8s-broker --brokercontext cluster1
> ```
> 
> To verify, we can save off the original `broker-info.subm` and compare with the recovered one.
> 
Thats what I tried to do with `[no_broker_found_error()](https://github.com/submariner-io/subctl/pull/467/files#diff-637916d64af30a6e6f9d31f415d8250201b56786188c35a41c2ee0763bcffb84R43)`. Dind't work as expected so I have changed the implementation.

> Expect failures:
> 
> ```
> _subctl recover-broker-info --namespace non-existent --context cluster1
> _subctl recover-broker-info --namespace submariner-k8s-broker --context cluster2
> _subctl recover-broker-info --brokernamespace non-existent --brokercontext cluster1
> _subctl recover-broker-info --brokernamespace submariner-k8s-broker --brokercontext cluster2
> ```
> 
> For these, verify a non-zero exit code.
> 
> After both clusters are uninstalled on L194, we can re-deploy the broker on cluster1 and test the case where there's no joined clusters and it generates a new IPsec secret:
> 
> ```
> _subctl recover-brokerinfo  --brokercontext cluster1
> _subctl recover-broker-info --namespace submariner-k8s-broker --context cluster1
> ```
> 
> For the broker-only install case, we can join cluster2:
> 
> `_subctl recover-broker-info --brokercontext cluster1 --context cluster2`

skitt commented on scripts/test/lib_subctl_recover_brokerinfo_tests.sh:
```diff
@@ -0,0 +1,133 @@
+          join_cluster cluster2
+        fi
+
+        use_context $context
```
Please don’t change the context during tests, in my experience it ends up creating confusion — especially since the context remains set for anything running after the tests.

It’s better to be explicit and specify the context for each `subctl` invocation.

skitt commented on scripts/test/lib_subctl_recover_brokerinfo_tests.sh:
```diff
@@ -0,0 +1,133 @@
+
+function no_broker_found_test() { # Blue
+  echo "Testing the scenarios where Broker is not installed on the specified context"
+  for cluster1subm in y n; do
+    for cluster2subm in y n; do
+      for context in cluster1 cluster2; do
```
I know I suggested the use of loops like this initially, but only if it helped clarify things. The end result here doesn’t benefit from the use of loops, at least not using the structure shown in the PR.

(As an example, it may ultimately be more useful to loop over an array of clusters; but the best way to go about this is to start with a sequential list of test commands, and see from that whether there’s opportunity and relevance in factoring code.)

skitt commented on pkg/cluster/info.go:
```diff
@@ -43,6 +44,7 @@ type Info struct {
 	Name             string
 	RestConfig       *rest.Config
 	ClientProducer   client.Producer
+	Broker           *v1alpha1.Broker
```
Structure fields should only be used if their contents have similar validity. When a function receives an `Info` struct, it knows that `Submariner` and `ServiceDiscovery` are fixed — either they’re present, or they’re not, but that’s not going to change, and it doesn’t need to do anything to have the value filled in. `Broker` is different; it’s only valid after `GetBroker` has been called. This means that it’s not a valid structure field; it should be dropped, and `GetBroker()` used whenever the Broker object is needed (with appropriate per-function local storage of results).

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+# Test subctl recover-broker-info invocations which will always fail (no matter the broker and Submariner installations and the current
+# context set) with broker not found error
+_subctl recover-broker-info --namespace non-existent && exit 1 #column 2
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster2 && exit 1 #column 4
```
I think `--namespace` extraneous here, ie it uses `--brokernamespace` with `-brokercontext`. Same with others below.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+_subctl recover-broker-info --namespace non-existent && exit 1 #column 2
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster2 && exit 1 #column 4
+_subctl recover-broker-info --context cluster2 && exit 1 #column 7
+_subctl recover-broker-info && exit 1 #5:1 - 8:1
```
This seems non-deterministic depending on which is the current context.... it seems we're assuming it's cluster2. Either way we should set the desired context.

Jaanki replied:
The first 3 commands would always error out with no broker found no matter the context.

tpantelis replied:
Ok but Im referring to this one: `_subctl recover-broker-info`. We don't set the context so it's non-deterministic. If you're adverse to setting the context then we should omit this case. It's effectively the same as `--context` and `--namespace` anyway.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster1 --context cluster2
+cmp broker-info.subm.orig broker-info.subm
+
+if [ "$(kubectl config current-context)" == "cluster1" ]; then # finishes row 4
```
Again this is non-deterministic. We should change the current context to cluster1.

Jaanki replied:
I have tried to avoid setting the context. And address all "test situations" rather than running all command variations.

Jaanki replied:
The test is written in such a way that whatever the context, evey situation is taken care of. Added context checks (if content = cluster1) wherever applicable.

tpantelis replied:
I'm not sure I follow but I don't prefer the context checks, again it's non-deterministic as it depends  on whatever prior code set the context to. And on L201 above you don't check the context which is inconsistent. Again if you don't want to make it deterministic then we should omit the case where we pass no args.

skitt replied:
It’s up to you to control what the context is; you shouldn’t ever need to check whether it’s `cluster1`, you should know that it _is_ `cluster1` (or isn’t). If you specifically want to test commands with no argument, that qualifies IMO as a valid reason to set the context, with a comment to that effect.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+
+# Test subctl recover-broker-info invocations with both Broker and Submariner installed  # remaining row 8
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster1
+cmp broker-info.subm.orig broker-info.subm
```
We should delete the _broker-info.subm_ for the next test to ensure each invocation actually generates the file.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
 _subctl cloud prepare generic --kubecontext cluster1 && exit 1
 
+# Rename old broker-info.subm to be able to compare it with the recovered broker-info.sum
+echo "Renaming broker-info.sum to broker-info.subm.orig"
```
```suggestion
echo "Renaming broker-info.subm to broker-info.subm.orig"
```

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+  _subctl recover-broker-info
+  cmp broker-info.subm.orig broker-info.subm
+fi
+
```
In addition to my comments above, I think some of the tests above are redundant. Also my preference is to do the positive tests first.

I suggest:

```
# Test subctl recover-broker-info invocations with both Broker and Submariner installed  on the same cluster
_subctl recover-broker-info --brokercontext cluster1
cmp ...
_subctl recover-broker-info --namespace submariner-k8s-broker --context cluster1
kubectl config use-context cluster1
subctl recover-broker-info

# Failure tests
# Invalid broker namespace
_subctl recover-broker-info --namespace non-existent --context cluster1 && exit 1
_subctl recover-broker-info --brokernamespace non-existent --brokercontext cluster1 && exit 1
# No broker installed
_subctl recover-broker-info --brokercontext cluster2
_subctl recover-broker-info --namespace submariner-k8s-broker --context cluster2 && exit 1
```

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+# context set) with broker not found error
+_subctl recover-broker-info --namespace non-existent && exit 1 #column 2
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster2 && exit 1 #column 4
+_subctl recover-broker-info --context cluster2 && exit 1 #column 7
```
What's the namespace used here? Do we default `--namespace` to _submariner-k8s-broker_?

Jaanki replied:
yes. Default namespace with `--context` and  `--brokercontext` is ` submariner-k8s-broker`. Both can be specified via `--namespace` and `--brokernamespace` flag respectively.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+cmp broker-info.subm.orig broker-info.subm
+
+# All the other commands fail
+_subctl recover-broker-info && exit 1
```
We've already tested no args above.

Jaanki replied:
a no args command will behave differently in different scenarios. I am testing how each command will behave in different scenarios.

tpantelis replied:
No args is effectively the same as `--context` and `--namespace` so it's redundant. Plus, as pointed out earlier, it's non-deterministic which context is current. Also on L209 above you do a context check but here you don't - it's inconsistent.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+
+# All the other commands fail
+_subctl recover-broker-info && exit 1
+_subctl recover-broker-info --namespace hfwiegld && exit 1
```
We've already tested invalid namespace above.

Jaanki replied:
yes. This can be deleted

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+_subctl recover-broker-info && exit 1
+_subctl recover-broker-info --namespace hfwiegld && exit 1
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster1 && exit 1
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster2 && exit 1
+_subctl recover-broker-info --context cluster2 && exit 1
```
We've already tested no broker installed above.

For this section I suggest:

```
# Test subctl recover-broker-info invocations with Submariner not installed on the Broker cluster
_subctl recover-broker-info --brokercontext cluster1 --context cluster2
cmp ...

# Failure tests
_subctl recover-broker-info --brokercontext cluster1 && exit 1
```

tpantelis commented on scripts/test/system.sh:
```diff
@@ -187,7 +188,50 @@ _subctl cloud prepare generic --context cluster1
+# Test subctl uninstall invocations
 _subctl uninstall -y --context cluster2
-_subctl uninstall -y --kubeconfig "${KUBECONFIGS_DIR}"/kind-config-cluster1
+
+# Test subctl recover-broker-info invocations with Submariner not installed on any cluster (Row 3, 7)
+_subctl recover-broker-info && exit 1
+_subctl recover-broker-info --namespace hfwiegld && exit 1
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster1 && exit 1
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster2 && exit 1
+_subctl recover-broker-info --context cluster2 && exit 1
+_subctl recover-broker-info --namespace submariner-k8s-broker --brokercontext cluster1 --context cluster2 && exit 1
```
These tests are really redundant with above, ie we've already tested the case where there's no broker installed.

Jaanki replied:
But this is the scenario where Broker is installed on cluster 1 but no Submariner is installed. I have still removed few commands that were actually covered before.

skitt commented on pkg/cluster/info.go:
```diff
@@ -220,3 +221,14 @@ func MergeImageOverrides(imageOverrides map[string]string, localImageOverrides [
 	return imageOverrides, nil
 }
+
+func (c *Info) GetBroker(namespace string) (*v1alpha1.Broker, error) {
```
The commit message no longer describes this change, please update it.

mkolesnik commented on pkg/cluster/info.go:
```diff
@@ -220,3 +221,14 @@ func MergeImageOverrides(imageOverrides map[string]string, localImageOverrides [
 	return imageOverrides, nil
 }
+
+func (c *Info) GetBroker(namespace string) (*v1alpha1.Broker, error) {
```
Why is this a separate commit?

There is no usage of this function here, nor any changes other than adding this function.

Usually commits need to be self contained, breaking them up like this just introduces needless complexity - e.g. looking at the history you dont see it with it's actual usage, and if you want to cherry pick the changes, you need to remember to cherry pick this commit as well.

Jaanki replied:
Because the original commit had different content and it has eveolved since then.

skitt replied:
As a result of those changes, I don’t think there’s any value in keeping the first commit separate — adding a function on its own isn’t useful, it only becomes fully understandable when the code which calls it is available. Please squash the two commits.

Jaanki replied:
Already merged.

skitt replied:
Could you push the version with the squashed commits then? The PR still shows two commits, the first of which only adds the `RecoverData()` function.

Jaanki replied:
I have merged the `GetBroker` function commit with the commit that adds `recover` command. `RecoverData()` function is still a different commit

skitt replied:
Yes, and my comment still stands: adding a function on its own isn’t useful, it only becomes fully understandable when the code which calls it is available. Think about it this way: if the second commit were reverted for any reason, the first commit would also be reverted (because `RecoverData()` wouldn’t have any users); the first commit has no justification without the second commit, so it shouldn’t be separate.

mkolesnik commented on pkg/broker/recover_broker_info.go:
```diff
@@ -0,0 +1,56 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func RecoverData(
```
Why an entirely new function?

Seems to me that you can refactor `broker.WriteInfoToFile` to suit your needs pretty easily.

It will also give us one place to maintain for writing broker files, instead of another that we need to keep track of.

Jaanki replied:
There are reasons for this:

1. `broker.WriteInfoToFile` everytime generates a new PSK secret which we don't want. It would need to be tweaked and **can't be used as is**. 
2. It expects components, customDomains as arguments. So to get these from Broker, a new function would anyways need to be written to fetch them first and then pass to `broker.WriteInfoToFile`.
3. It expects `ipsec File` (file that contains PSK Secret specified by the user) as an argument. We are not providing this support via `recover-broker-info` command
4. `recover-broker-info` is a new command and it makes sense that it goes to its dedicated function call

I have reused `data.writeToFile` that `broker.WriteInfoToFile` also uses to write the collected data to `broker-info.subm`

mkolesnik replied:
You missed the point, if you use a single function and feed it the necessary parameters, should the broker info change then both callers will be either doing the same thing or would both need to be updated.

If you use seperate functions, the changes need to be remembered to be made in both by devs (which i can guarantee wont happen)

> I have reused data.writeToFile that broker.WriteInfoToFile also uses to write the collected data to broker-info.subm

That just outputs the info to file, nothing fancy

skitt replied:
See https://github.com/submariner-io/subctl/pull/588.

mkolesnik replied:
Should we fold it into this one, or treat as standalone PR?

Personally, I think squashing would be better.

skitt replied:
I’d rather keep them separate, let’s get this one merged (once the remaining comments are addressed) and finish reviewing #588 separately.

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+	Use:   "recover-broker-info",
+	Short: "Recovers the broker-info.subm file from the installed Broker",
+	Run: func(cmd *cobra.Command, args []string) {
+		exit.OnError(recoverBroker(cli.NewReporter()))
```
Would be more readable if the high level `if` was here, and the calls to `recoverRestConfigProducer.RunOnSelectedPrefixedContext` and `recoverRestConfigProducer.RunOnSelectedContext` were named functions called from here.

e.g. something like
```suggestion
		status := cli.NewReporter()

		// if --brokerconfig flag provided, get the broker and proceed to get Submariner
		contextFound, err := recoverRestConfigProducer.RunOnSelectedPrefixedContext("broker",
			recoverBrokerFromConfigContext, status)
		exit.OnError(err)

		// if --brokerconfig not provided, search for broker on current context and proceed to get Submariner
		if !contextFound && err == nil {
			err = recoverRestConfigProducer.RunOnSelectedContext(recoverBrokerFromCurrentContext, status)
		}

		exit.OnError(err)
```

Jaanki replied:
comments help with readability IMO

mkolesnik replied:
I'm not against the comments, I'm referring to the high level of indentation happening inside the `recoverBroker` function.

If you use named functions and call them from here, instead of this monolithic function, you lower the level of indentation and make the code much more readable.

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -127,8 +128,8 @@ subm_ns=submariner-operator
-deploy_env_once
-
+#deploy_env_once
+<<comment
```
?

Jaanki replied:
This is a left over of my local testing setup. Its been removed.

mkolesnik commented on scripts/test/system.sh:
```diff
@@ -186,8 +187,50 @@ _subctl benchmark throughput "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECO
+
+# All the tests assume that the current context is cluster2
+# Please refer the test matrix sheet for better understanding
+# https://docs.google.com/spreadsheets/d/1a4Ufw1EqBcms5FfDPx9zK_ep7zw77__OvCWyr_M6jJc/edit#gid=0
```
This is not publicly accessible, if you want to keep the link here it should be

Jaanki replied:
Wil check it

tpantelis replied:
Please remove the link - I don't think a spreadsheet is necessary as the test cases have been reduced and simplified and should be intuitive looking at the code.

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+	"github.com/submariner-io/subctl/pkg/cluster"
+)
+
+var recoverRestConfigProducer = restconfig.NewProducer().WithPrefixedContext("broker").
```
I find the whole UX of this confusing, since you have `--context`, `--namespace`, `--brokercontext` and `--brokernamespace`.

As a user I'm supposed to provide both `--context` and `--brokercontext` (unless the broker happens to be on the same cluster).
I would expect to provide, as a user, only the `--context` and the command should figure out whatever details (technically only components and custom domains right now) from that cluster's connection to the broker.

I was thinking the idea was to be able to recover the info given only the broker cluster, but that isn't possible and makes the whole idea of "just specify `--brokercontext`" not really usable.
Seems that we have to get details from an installed Submariner instance anyhow.

I think we should only have `--context` flag.
If submariner is installed on the cluster, the command can then figure out most of the details from the `Submariner` CR and get any additional info by connecting to the broker from these details and querying it directly.

This will be a much simpler and friendlier UX, and avoids any confusion or "invalid combinations" that can bet sent by the user.

Jaanki replied:
The idea is to provide a smart interaction for the user. If there is no context specified (neither `--context` nor `--brokercontext`), then you can still recover the file. But if the user wants to specify either the Broker and Submariner clusters, this implementation allows that and errors out if the specified contexts are not correct. So this implementation doesn't force user to specify any context and the code sends out appropriate messages if something is wrong.

Also, what you are suggesting is a design change. The right time for it would have been during the design phase.

mkolesnik replied:
> The idea is to provide a smart interaction for the user. If there is no context specified (neither `--context` nor `--brokercontext`), then you can still recover the file. But if the user wants to specify either the Broker and Submariner clusters, this implementation allows that and errors out if the specified contexts are not correct. So this implementation doesn't force user to specify any context and the code sends out appropriate messages if something is wrong.

I don't understand your point.

The user must point (either explicitly or explicitly) to a cluster that has Submariner installed on it, right?
Why confuse them with a need to point to the cluster where the broker is, if the info is already available.

> 
> Also, what you are suggesting is a design change. The right time for it would have been during the design phase.

So you're saying that a worse design should be honored since it was agreed upon, instead of a better one?
That's not quite agile..

tpantelis replied:
It would be ideal if we could get all the info via the `Submariner` resource and not deal with a broker context. In order to do that we'd need to use the broker URL and credentials from the `Submariner` resource to retrieve the _submariner-k8s-broker-admin_  `Secret` and the Broker resource from the broker namespace.  But I'm not sure if the credentials have the permissions to do so. I assume that's why we're requiring a broker context from the user.

mkolesnik replied:
What are those credentials? I can't seem to find it in the RBAC code to figure out if it has the correct permissions

mkolesnik replied:
OK found it in https://github.com/submariner-io/subctl/blob/devel/pkg/broker/rbac.go#L51

So we could add a rule there to allow getting/listing brokers (which shouldn't be a problem as the broker doesn't have any secret data anyway, and the role is namespaced so the SA can only list the broker in the relevant namespace) and then it should be possible AFAICT.

Then we'd have a much simpler and less confusing UX

tpantelis replied:
I assume we'd have to build a `rest.Config` from  `Submariner.Spec.BrokerK8sApiServer`, `Submariner.Spec.BrokerK8sApiServerToken` et al but I **think** the associated RBAC is defined [here](https://github.com/submariner-io/subctl/blob/92115ef06fc50134c0f37cf3048afc6b64364911/pkg/broker/rbac.go#L82).

skitt replied:
At some point the token will be going away.

tpantelis replied:
I believe it's https://github.com/submariner-io/subctl/blob/92115ef06fc50134c0f37cf3048afc6b64364911/pkg/broker/rbac.go#L82 from which `Submariner.Spec.BrokerK8sApiServerToken` is derived. So in addition to getting the `Broker` resource, it would also need "get" access to `ServiceAccount` and `Secret`. This doesn't seem like an issue. @skitt WDYT?

tpantelis replied:
> At some point the token will be going away.

I assume the `Submariner` resource will still contain the info needed to connect to the broker cluster. Whatever form that is, I assume we should be able to use that here assuming we setup the proper RBAC.

skitt replied:
Yes, if joined clusters can connect to the broker, it would be possible to retrieve whatever information is needed to connect from somewhere else. (That may not always be true, but it is true for now.)

I’m under the impression we discussed this during the enhancement proposal review for the feature, but I can’t remember why we eschewed retrieving the broker credentials from a connected cluster instead of asking the user for broker credentials. Even if the user’s system can’t access the broker for some reason, the user could run `subctl` in a pod on a connected cluster.

tpantelis replied:
> but I can’t remember why we eschewed retrieving the broker credentials from a connected cluster

yeah I don't know why either.

mkolesnik replied:
Seems the original discussion was happening here - https://github.com/submariner-io/enhancements/pull/144#discussion_r1050265381

But I think the original assessment by @tpantelis is either wrong, or I'm misunderstanding what we need to do.

To me it seems that the only access needed to be added to the default `Role: submariner-k8s-broker-cluster` is to be able to read a `Broker` resource (to get the `components` and `defaultCustomDomains` from it).

Am I missing something?

skitt replied:
Given a context capable of accessing the `submariner-operator` repository on a joined cluster, we can retrieve the broker secret named in the `Submariner` CR, and construct a client using that to access the broker using the appropriate credentials for that operator. The broker credentials don’t need to give access to a role which is itself capable of reading secrets.

skitt replied:
May I suggest the following compromise with regards to this PR? Given that there already is a working implementation using separate contexts, could we consider merging that (once other pending review comments have been resolved), and work on simplifying the UX separately?

mkolesnik replied:
Generally I'm not a fan of this mentality of "lets merge this even if we don't agree and fix it later in a separate PR".

* We still need to review this "less optimal solution", which is harder to review IMHO because of all this confusion.
* It still gets written in the history forever, and in a segmented fashion which makes browsing the history a bit more complicated.
* There's a risk that later may not actually happen.

If this was some unrelated change, merging now and adding or fixing something else later makes more sense, but as it touches on the PR contents themselves  IMHO this isn't a good compromise.

But then again I'm not a project committer/owner so eventually its up to committers/owners to decide.

skitt replied:
Well, another way to see this is that it implements [the proposed enhancement](https://github.com/submariner-io/enhancements/blob/devel/subctl/issue-89-reconstructing-broker-info.md), and in the context of that EP I don’t think there’s any confusion here. Strictly speaking, the EP should be updated if we want to pursue your suggested approach.

I take it “a segmented fashion” refers to the use of multiple commits; that would be addressed in the review.

The risk that later may not actually happen also applies here — requesting changes may result in the changes being implemented, or it may result in the PR being abandoned; there’s no guarantee. (Basically, if this is important to the project, it will happen, whether in this PR or another; if not, it won’t, whether in this PR or another.)

mkolesnik replied:
> Well, another way to see this is that it implements [the proposed enhancement](https://github.com/submariner-io/enhancements/blob/devel/subctl/issue-89-reconstructing-broker-info.md), and in the context of that EP I don’t think there’s any confusion here. Strictly speaking, the EP should be updated if we want to pursue your suggested approach.

I already replied to this talking point here - https://github.com/submariner-io/subctl/pull/467#discussion_r1105879456

skitt replied:
> > Well, another way to see this is that it implements [the proposed enhancement](https://github.com/submariner-io/enhancements/blob/devel/subctl/issue-89-reconstructing-broker-info.md), and in the context of that EP I don’t think there’s any confusion here. Strictly speaking, the EP should be updated if we want to pursue your suggested approach.
> 
> I already replied to this talking point here - [#467 (comment)](https://github.com/submariner-io/subctl/pull/467#discussion_r1105879456)

No, you replied to a different point. I’m not saying this should be applied because it implements the proposed enhancement. I’m saying that the enhancement proposal makes this PR easy to understand, and that if changes are to be made, for consistency the EP should be updated too.

tpantelis replied:
The consensus is to finish this PR as is since it does follow the previously agreed upon proposal. Subsequently the feasibility of simplifying the UX by removing the `brokercontext` arg will be explored and the enhancement proposal modified accordingly followed by the implementation.

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+
+func tryToRecoverFromBroker(brokerCluster *cluster.Info, brokerObj *v1alpha1.Broker,
+	brokerNamespace string, clusters []submarinerv1.Cluster, status reporter.Interface) (bool, error) {
+	status.Start("Checking if there are any cluster(s) joined to the Broker")
```
```suggestion
	status.Start("Checking if there are any Submariner clusters joined to the Broker")
```

Jaanki replied:
I think `submariner` is implicit here. If there is no Submariner installed, the cluster won't be listed.

tpantelis replied:
OK but I think it should be "_clusters_" and not "_cluster(s)_".

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+		return false, status.Error(
+			errors.New(
+				"no Submariner is joined to the Broker. "+
+					"Re-run broker-deploy command to regenerate broker-info.subm file"), "")
```
```suggestion
				"no clusters are joined to the Broker. "+
					"Please re-run the 'deploy-broker' command to regenerate the broker-info.subm file"), "")
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+	status.Success("Found cluster(s) joined to the Broker")
+	status.End()
+
+	status.Start("Checking if Submariner is installed on the same cluster as Broker")
+    defer status.End()
```
We don't really need to start a new status here.

```suggestion
```

Jaanki replied:
its like logical differentiating between different stepsthat are been taken.

tpantelis replied:
To me, the reason for using `Start` is to get the spinner if it takes too long, specifically if an operation makes a remote call that may arbitrarily hang for a while. That's not the case here.  The end result indicates the step, ie "Found Submariner installed..." or "Submariner not installed..." so the "Checking ..." statement is extraneous. The less output to convey the point, the better.

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+    defer status.End()
+
+	if isSubmJoinedToBroker(clusters, brokerCluster) {
+		status.Success("Found Submariner installed on the same cluster")
```
```suggestion
		status.Success("Found a local Submariner installation joined to the Broker")
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+			return recoverRestConfigProducer.RunOnSelectedContext(
+				func(submCluster *cluster.Info, namespace string, status reporter.Interface) error {
+					if isSubmJoinedToBroker(clusters, submCluster) {
+						status.Success("Found Submariner on cluster %s", submCluster.Name)
```
```suggestion
						status.Success("Found a Submariner installation on cluster %q  joined to the Broker", submCluster.Name)
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+					}
+
+					return status.Error(
+						errors.New(fmt.Sprintf("submariner is not installed on this cluster %s. "+
```
```suggestion
						errors.New(fmt.Sprintf("submariner is not installed on cluster %q. "+
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+		}, status)
+
+	// if --brokerconfig not provided, search for broker on current context and proceed to get Submariner
+	if !contextFound && err == nil {
```
Again, return early here to reduce indentation.

```suggestion
        if contextFound || err != nil {
            return err
        }
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+				}
+
+				return status.Error(
+					errors.New(fmt.Sprintf("submariner is not installed on this cluster %s. "+
```
```suggestion
					errors.New(fmt.Sprintf("submariner is not installed on cluster %q. "+
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+func getBroker(clusterInfo *cluster.Info, namespace, notFoundMsg string, status reporter.Interface) (*v1alpha1.Broker, error) {
+	status.Start("Checking if Broker is installed on the cluster %q in namespace %q", clusterInfo.Name, namespace)
```
```suggestion
	status.Start("Checking if the Broker is installed on cluster %q in namespace %q", clusterInfo.Name, namespace)
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,172 @@
+		return nil, status.Error(err, "")
+	}
+
+	status.Success("Broker found")
```
Not necessary.

```suggestion
```

Jaanki replied:
Its a verbal output. And maintains the flow. L122 saying "checking if found or not" and this line replies to it.

tpantelis replied:
IMO this L122 doesn't really need an explicit reply, ie the green checkmark is the positive reply and the fact that it proceeds. If it fails then we get a negative message in reply.  Plus this message is rather terse and doesn't provide any more information as a reply to the start message. 

OTOH, if the message was more general like "_Checking if the Broker is installed_", then a reply like "_Found the Broker installed on cluster "foo" in namespace "bar"_" is useful b/c it provides additional information. Note: I'm not saying we should change the message like that, just using it as an example of when an explicit reply is useful.

tpantelis commented on pkg/cluster/info.go:
```diff
@@ -220,3 +221,25 @@ func MergeImageOverrides(imageOverrides map[string]string, localImageOverrides [
+	return broker, err
+}
+
+func (c *Info) GetJoinedClusters(namespace string) ([]submarinerv1.Cluster, error) {
```
```suggestion
func (c *Info) GetClusters(namespace string) ([]submarinerv1.Cluster, error) {
```

tpantelis commented on scripts/test/system.sh:
```diff
@@ -186,8 +187,50 @@ _subctl benchmark throughput "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECO
+echo "Renaming broker-info.subm to broker-info.subm.orig"
+mv "${DAPPER_SOURCE}"/output/broker-info.subm "${DAPPER_SOURCE}"/output/broker-info.subm.orig
+
+# All the tests assume that the current context is cluster2
```
This statement is stale - we're no longer making assumptions about the current context.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -186,8 +187,50 @@ _subctl benchmark throughput "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECO
+_subctl recover-broker-info --brokercontext cluster2 && exit 1
+_subctl recover-broker-info --context cluster2 && exit 1
+
+# Test subctl recover-broker-info invocations with Submariner installed on both or just Broker cluster
```
"just Broker cluster" isn't accurate.

```suggestion
# Test subctl recover-broker-info invocations with both Broker and Submariner installed  on the same cluster
```

tpantelis commented on scripts/test/system.sh:
```diff
@@ -186,8 +187,50 @@ _subctl benchmark throughput "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECO
+cmp "${DAPPER_SOURCE}"/output/broker-info.subm.orig broker-info.subm
+rm -f broker-info.subm
+
+_subctl recover-broker-info --brokercontext cluster1 --context cluster2
+cmp "${DAPPER_SOURCE}"/output/broker-info.subm.orig broker-info.subm
+rm -f broker-info.subm
```
We can remove this case as the test args are redundant with L223 below but the behavior is actually the same as L207 above. Since Submariner is installed on the Broker, the `--context` are isn't used.

tpantelis commented on scripts/test/system.sh:
```diff
@@ -186,8 +187,50 @@ _subctl benchmark throughput "${KUBECONFIGS_DIR}"/kind-config-cluster1 "${KUBECO
+# Please refer the test matrix sheet for better understanding
+# https://docs.google.com/spreadsheets/d/1a4Ufw1EqBcms5FfDPx9zK_ep7zw77__OvCWyr_M6jJc/edit#gid=0
+
+# Test subctl recover-broker-info invocations which will always fail (no matter the broker and Submariner installations and the current
+# context set) with broker not found error # all blues
+_subctl recover-broker-info --namespace non-existent && exit 1
+_subctl recover-broker-info --brokercontext cluster2 && exit 1
+_subctl recover-broker-info --context cluster2 && exit 1
```
Also please move the negative test cases after the positive cases below L217.

```suggestion
# Failure tests
# Invalid broker namespace
_subctl recover-broker-info --namespace non-existent --context cluster1 && exit 1
_subctl recover-broker-info --brokernamespace non-existent --brokercontext cluster1 && exit 1
# No broker installed
_subctl recover-broker-info --brokercontext cluster2
_subctl recover-broker-info --context cluster2 && exit 1
```

tpantelis commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,171 @@
+					"subm file"), "")
+	}
+
+	status.Success("Found cluster(s) joined to the Broker")
```
Similar to a comment elsewhere, I think the green checkmark is a sufficient positive reply unless we provide more info in the message.  So I think either remove this message or change it to:

```suggestion
	status.Success("Found %d cluster(s) joined to the Broker", len(clusters))
```

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,175 @@
+	return recoverRestConfigProducer.RunOnSelectedContext(
+		func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error {
+			brokerObj, err := getBroker(clusterInfo, namespace, "Please specify the cluster where the Broker is installed "+
+				"via the "+"`--brokerconfig` or `--brokercontext` flag and the namespace via the --brokernamespace flag", status)
```
```suggestion
				"via the `--brokerconfig` or `--brokercontext` flag and the namespace via the `--brokernamespace` flag", status)
```

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,175 @@
+				return err
+			}
+
+			clusters, err := clusterInfo.GetClusters(namespace)
+			if err != nil {
+				return status.Error(err, "error listing joined clusters")
+			}
+
+			ok, err := tryToRecoverFromBroker(clusterInfo, brokerObj, namespace, clusters, status)
+			if ok || err != nil {
+				return err
+			}
```
```suggestion
			return broker.RecoverData(clusterInfo, clusterInfo, brokerObj, clusterInfo.Submariner.Spec.BrokerK8sRemoteNamespace, status)
```

Shouldn't you just call `broker.RecoverData` directly in such a case?
You already have the cluster info in hand, and either the broker is on the same cluster or you should've exited before, so just call it to recover the info

skitt replied:
As I understand it, this handles scenarios where the joined cluster being processed isn’t connected to the broker being processed. It’s unlikely but possible, and the result would be an incorrect `broker-info.subm`.

mkolesnik replied:
> As I understand it, this handles scenarios where the joined cluster being processed isn’t connected to the broker being processed. It’s unlikely but possible, and the result would be an incorrect broker-info.subm.

AFAICT it would only produce wrong broker info file if submariner is installed on the cluster and connected to a different broker with the same namespace as a broker that is installed locally (but is not connected to the local cluster).

I guess this could be a valid though weird scenario, as resources are synced to the n/s only on the broker so it shouldnt cause problems in theory.

skitt replied:
We’ve seen exactly this scenario in the past, when users re-deployed the broker on a different cluster and then forgot which cluster was connected to which broker.

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,175 @@
+				fmt.Errorf("submariner is not installed on cluster %q. "+
+					"Please specify the cluster where Submariner is installed via `--kubeconfig` or `--context` flag"+
+					"", clusterInfo.Name), "")
+		}, status)
```
```suggestion
			if clusterInfo.Submariner == nil {
				return fmt.Errorf("submariner is not installed on cluster %q. "+
					"Please specify a cluster where Submariner is installed via `--kubeconfig` or `--context` flag"+
					"", clusterInfo.Name)
			}

			return broker.RecoverData(clusterInfo, clusterInfo, brokerObj, clusterInfo.Submariner.Spec.BrokerK8sRemoteNamespace, status)
```

Shouldn't you just call `broker.RecoverData` directly in such a case?
You already have the cluster info in hand, and either the broker is on the same cluster or you should've exited before, so just call it to recover the info

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,175 @@
+
+	brokerObj, err := clusterInfo.GetBroker(namespace)
+	if apierrors.IsNotFound(err) {
+		return nil, status.Error(fmt.Errorf("the Broker is not installed on the specified cluster. %s", notFoundMsg), "")
```
You should also specify namespace as it could be significant for the failure info

mkolesnik commented on pkg/cluster/info.go:
```diff
@@ -220,3 +221,25 @@ func MergeImageOverrides(imageOverrides map[string]string, localImageOverrides [
 	return imageOverrides, nil
 }
+
+func (c *Info) GetBroker(namespace string) (*v1alpha1.Broker, error) {
+	broker := &v1alpha1.Broker{}
+	err := c.ClientProducer.ForGeneral().Get(
+		context.TODO(), controllerClient.ObjectKey{
+			Namespace: namespace,
+			Name:      brokercr.Name,
+		}, broker)
+
+	return broker, errors.Wrap(err, "error retrieving Broker")
+}
+
+func (c *Info) GetClusters(namespace string) ([]submarinerv1.Cluster, error) {
+	clusters := &submarinerv1.ClusterList{}
+
+	err := c.ClientProducer.ForGeneral().List(context.TODO(), clusters, controllerClient.InNamespace(namespace))
+	if err != nil {
+		return nil, errors.Wrap(err, "error retrieving Clusters")
+	}
+
+	return clusters.Items, nil
+}
```
These should probably be declared after `OperatorNamespace` as they're also instance functions.

mkolesnik reviewed (changes requested):
Some comments

mkolesnik commented:
The commit message doesn't make much sense right now:

```
Add recover-brokerinfo command

and add system tests

Epic: https://github.com/submariner-io/enhancements/issues/143

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

Get data and populate broker-info.subm

Epic: https://github.com/submariner-io/enhancements/issues/143

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>
```

mkolesnik commented on cmd/subctl/recover_broker_info.go:
```diff
@@ -0,0 +1,176 @@
+		exit.OnError(err)
+
+		// if --brokerconfig not provided, search for broker on current context and proceed to get Submariner
+		if !contextFound && err == nil {
```
nit: I guess `exit.OnError(err)` on L#51 will make this check here unneeded.

```suggestion
		if !contextFound {
```

mkolesnik reviewed (approved):
Please make sure to update design and remove the `--broker` flags

dfarrell07 commented:
Release notes are here: https://github.com/submariner-io/subctl/pull/630
